File: src/main/java/io/github/legacy_fabric_community/serialization/api/NbtConfigManager.java
package io.github.legacy_fabric_community.serialization.api;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Path;

import javax.annotation.Nonnull;

import io.github.legacy_fabric_community.serialization.nbt.NbtOps;
import com.mojang.serialization.Codec;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtIo;

public class NbtConfigManager<T> extends ConfigManager<T> {
    @Nonnull
    private final T defaultValue;

    protected NbtConfigManager(Path configPath, Codec<T> codec, @Nonnull T defaultValue) {
        super(configPath, codec);
        this.defaultValue = defaultValue;
    }

    @Override
    public void serialize() throws IOException {
        NbtIo.writeCompressed(
                (CompoundTag) this.codec.encodeStart(
                        NbtOps.INSTANCE,
                        this.config
                ).getOrThrow(false,
                        PRINT_TO_STDERR
                ),
                new FileOutputStream(this.configPath.toFile())
        );
    }

    @Override
    public void deserialize() throws IOException {
        this.config = this.codec.decode(
                NbtOps.INSTANCE,
                NbtIo.readCompressed(
                        new FileInputStream(this.configPath.toFile())
                )
        ).getOrThrow(
                false,
                PRINT_TO_STDERR
        ).getFirst();
    }

    @Override
    protected void writeDefaultData() throws IOException {
        NbtIo.writeCompressed(
                (CompoundTag) this.codec.encodeStart(
                        NbtOps.INSTANCE,
                        this.defaultValue
                ).getOrThrow(false,
                        PRINT_TO_STDERR
                ),
                new FileOutputStream(this.configPath.toFile())
        );
    }

    @Nonnull
    public T getDefaultValue() {
        return this.defaultValue;
    }
}

File: shadowed-config-configmanagers/src/main/scala/io/github/legacy_fabric_community/shadowedconfig/configmanagers/NbtConfigManager.scala
package io.github.legacy_fabric_community.shadowedconfig.configmanagers

import java.io.IOException
import java.nio.file.{Files, Path}
import java.util.Optional
import java.util.function.Consumer

import com.mojang.serialization.Codec
import io.github.legacy_fabric_community.shadowedconfig.nbt.NbtOps
import javax.annotation.Nonnull
import net.minecraft.nbt.{CompoundTag, NbtIo}

object NbtConfigManager {
	protected val PRINT_TO_STDERR: Consumer[String] = (str: String) => {
		println(str)
	}
}

class NbtConfigManager[T] (override val configPath: Path, override val codec: Codec[T], @Nonnull val defaultValue: T) extends ConfigManager[T](configPath, codec) {
	@throws[IOException]
	override def serialize(): Unit = {
		NbtIo.writeCompressed(this.codec.encodeStart(NbtOps.INSTANCE, this.getConfig).getOrThrow(false, NbtConfigManager.PRINT_TO_STDERR).asInstanceOf[CompoundTag], Files.newOutputStream(this.configPath))
	}

	@throws[IOException]
	override def deserialize(): Unit = {
		this.config = (Optional.of(this.codec.parse(NbtOps.INSTANCE, NbtIo.readCompressed(Files.newInputStream(this.configPath))).getOrThrow(false, NbtConfigManager.PRINT_TO_STDERR)))
	}

	@throws[IOException]
	override def writeDefaultData(): Unit = {
		NbtIo.writeCompressed(this.codec.encodeStart(NbtOps.INSTANCE, this.defaultValue).getOrThrow(false, NbtConfigManager.PRINT_TO_STDERR).asInstanceOf[CompoundTag], Files.newOutputStream(this.configPath))
	}

	@Nonnull def getDefaultValue: T = this.defaultValue
}
